replaced git-standup with gitlog

mojoaxel opened the pull request:
I replaced [git-standup](https://www.npmjs.com/package/git-standup) with [gitlog](https://www.npmjs.com/package/gitlog). 
This should solve issues #11 and #14 :tada: 
This is an alternative to #8.

mojoaxel commented:
Maybe this also fixes #17

madve2 commented:
The nice thing about this is that it also works on Windows! Yay, thank you!

notwaldorf commented on package.json:
```diff
@@ -4,6 +4,9 @@
   "description": "A terminal that tries to take care of you 💖",
   "main": "care.js",
   "author": "Monica Dinculescu <noms@google.com>",
+  "contributors": [
```
I haven't been doing this for anyone else who has been sending PRs, so I'm inclined to just have the GitHub contributors reflect this. WDYT?

mojoaxel replied:
No problem at all! Let's keep the package.json slim :wink:

notwaldorf commented:
I tried running your patch locally, but I get:
```
Error: Repo location does not exist
    at gitlog (/Users/noms/Code/tiny-terminal-code/node_modules/gitlog/index.js:59:11)
    at async.each (/Users/noms/Code/tiny-terminal-code/care.js:124:5)
    at /Users/noms/Code/tiny-terminal-code/node_modules/async/dist/async.js:3047:16
```

notwaldorf commented:
Ohhh I see. So the advantage of `git-standup` is that you can just pass a directory that contains a bunch of other repos (like `~/Code` contains all my projects), and it would just crawl through them. With this approach, it looks like I have to list every repo? I have about 140, so I don't think that would work 😂

mojoaxel commented:
>  With this approach, it looks like I have to list every repo? I have about 140, so I don't think that would work

Good, point! I see if I improve further... :metal:

mojoaxel commented:
@notwaldorf  I introduced a new env var `TTC_REPOS_DEPTH` to define the max directory-depth.

mojoaxel commented on package.json:
```diff
@@ -13,9 +13,12 @@
+    "is-git": "0.0.1",
     "parrotsay-api": "^0.1.1",
     "scraperjs": "^1.2.0",
+    "subdirs": "github:mojoaxel/subdirs",
```
This is not perfect. I'm waiting for https://github.com/jarofghosts/subdirs/pull/2 to get merged...

mojoaxel commented:
@notwaldorf Ok, I now also updated the README and introduced some error handling :sparkles:.

Now you have to decide if you want to get rid of the script-based `git-standup` solutions in favor of this pure-js implementation. If you want to stick with `git-standup` that's also OK with me, just close this PR than (no hard feelings).

madve2 commented:
I really hope this will be the solution going forward, because the git-standup script version will never work on Windows.

Garbee commented:
If Windows support with the helper script is needed that is certainly simple enough to be ported into a Windows script that can be executed based on the environment. No need to convert the entire subsystem over just for that.

madve2 commented:
Are you sure about that? Replacing a dependency with a similar one (e.g. git-standup => gitlog) is hardly "convert[ing] the entire subsystem", and there are clear benefits in having a platform-independent, pure JS solution over having platform specific scripts written for different environments. (That's the whole point of node.js after all.) Yes, PowerShell is perfectly capable of what the current shell script is doing, but is a ps1 script really worth writing and maintaining? Especially considering the burden of having to update and test both scripts from now on for the rest of the project's lifetime, which won't be easy as it will require testing on at least two different operating systems.

Garbee commented:
I never said it was a bad change, I simply was stating if Windows support is desired then this effort of changing the way the git logs are retrieved (the subsystem) is not entirely required. The script is simple enough to do in a native Windows shell to at least get some initial support until this solution can be more thoroughly looked over. From some experience in the Firefox's new HTML-based Debugger for their DevTools, handling git stuff well and in a compatible way through Node isn't always straightforward. Some care should be taken before rushing into it and then having quirky setups cause issues that need to be solved that otherwise aren't issues.

Garbee commented on care.js:
```diff
@@ -103,21 +108,83 @@ function doTheTweets() {
+ * and look for repositories.
+ * Calls `callback` with array of repositories.
+ */
+function getGitRepos(repos, depth, callback) {
```
It looks like PR #36 may land opening up more ES6 usage. It may be good to wait for that to land then refactor these extra functions into isolated modules to keep the main care file smaller.

notwaldorf commented:
Just running this straight out of the box gives me: 
<img width="676" alt="screen shot 2017-04-26 at 10 50 55" src="https://cloud.githubusercontent.com/assets/1369170/25448722/42cc2d56-2a6e-11e7-842c-dd8acc63b6e8.png">

mojoaxel commented:
@notwaldorf This is probably because of relative paths. You are right, this should work...I'll look into it and do some cleanup...

mojoaxel commented:
@notwaldorf Please check again, if you find the time.
I separated the git functions into their own module and fixed the issue with relative-paths.

notwaldorf commented:
Yayayay! Thanks for sticking through with this! It works great!

Here are some generic comments I have (I'll do a proper review of the code in a second):
- (this is probably an easy fix) but I no longer get the listing of which repos the commits belong to. I just get a long list of commits.
- since it's now been separated in a separate file, can the "new way" be an opt-in thing (maybe TTC_GIT=`gitstandup | gitlog`? So only the people who don't have `git-standup` working at all can try this, until we get more testing in? I'm worried a little about switching to a new library and then having a whole new set of issues to look at that

notwaldorf commented on care.js:
```diff
@@ -1,15 +1,18 @@
 var bunnySay = require('sign-bunny');
 var weather = require('weather-js');
 
+console.dir(gitbot);
```
remove :)

mojoaxel replied:
:scream: sorry...gone

notwaldorf commented on care.js:
```diff
@@ -1,15 +1,18 @@
 #!/usr/bin/env node
 var config = require(__dirname + '/config.js');
 var twitterbot = require(__dirname + '/twitterbot.js');
+var gitbot = require(__dirname + '/gitbot.js');
```
🙌

notwaldorf commented on care.js:
```diff
@@ -115,21 +118,35 @@ function doTheTweets() {
+  }
+
+  gitbot.findGitRepos(config.repos, config.reposDepth-1, (err, allRepos) => {
+    if (err) return showError(err);
```
`return` on a new line, please (here and everywhere)

mojoaxel replied:
noproblemo

notwaldorf commented on care.js:
```diff
@@ -115,21 +118,35 @@ function doTheTweets() {
+  }
+
+  function showError(err, box) {
+    getCommits(`😥  ${err}`, box);
```
Is this bit actually needed? For an error, can't we just do `box.content = err`?

mojoaxel replied:
good idea!

notwaldorf commented on config.js:
```diff
@@ -11,7 +11,10 @@ var config = {
+  repos: (process.env.TTC_REPOS || '~/Code').split(','),
+
+  // The directory-depth on how to look for git-repos. Use with care!
+  reposDepth: process.env.TTC_REPOS_DEPTH || 1,
```
this is already called `depth`, below

mojoaxel replied:
sorry...merges and stuff...will fix

mojoaxel commented:
I'll add the repos to the commits and think about a nice way for the transition!

mojoaxel commented:
@notwaldorf I added the repo names. I also added a var 'TTC_GITBOT' to switch between `git-standup` and the node implementation. Hope this works for you!

mojoaxel commented:
Interesting: `git-standup` seems to filter out merge-commits like e.g. this one: `Merge branch 'master' into gitlog`.
So both implementations do not result in the same 10 commits at the moment. Not sure if that's a bug or a feature :confused:

mojoaxel commented:
Update: I now also filter pure administrative commits like merges (gitlog has a attribute for that).

notwaldorf commented on care.js:
```diff
@@ -116,21 +117,54 @@ function doTheCodes() {
+  if (config.gitbot.toLowerCase() === 'gitstandup') {
+    todayBox.content = '';
+    config.repos.forEach(repo => {
+      var today = spawn(`cd ${repo} && git-standup`, [`-m ${config.depth}`, '-d 1'], {shell:true});
```
uhhhhhh so I'm not 100% sure this works, because git-standup is on the wrong path (and works locally, but doesn't work when installed as a package). Can you just revert this, and leave the old git-standup behaviour, so that this PR _just_ adds gitlog, and not other things? That way it's easier if something goes wrong after this PR :)

notwaldorf commented on care.js:
```diff
@@ -116,21 +117,54 @@ function doTheCodes() {
+    });
+    weekBox.content = '';
+    config.repos.forEach(repo => {
+      var today = spawn(`cd ${repo} && git-standup`, [`-m ${config.depth}`, '-d 7'], {shell:true});
```
same here

notwaldorf commented on care.js:
```diff
@@ -116,21 +117,54 @@ function doTheCodes() {
+  } else {
+    gitbot.findGitRepos(config.repos, config.depth-1, (err, allRepos) => {
+      if (err)
+        return todayBox.content = err;
```
this won't call `screen.render()` anywhere, will it?

mojoaxel replied:
good point! I'll improve..

notwaldorf commented on config.js:
```diff
@@ -11,11 +11,14 @@ var config = {
   depth: (process.env.TTC_REPOS_DEPTH || 1),
 
+  // Which method is to be used to read the git commits ('gitstandup' | 'gitlog').
+  gitbot: (process.env.TTC_GITBOT || 'gitstandup'),
```
👍 Awesome!

mojoaxel replied:
git-standup is the default settings for now. Let's give the gitlog implementation a try on different platforms, maybe we can remove the whole bash-scripting stuff all together at some point!

notwaldorf commented on package.json:
```diff
@@ -13,15 +13,17 @@
     "twit": "^2.2.5",
     "weather-js": "^2.0.0"
   },
-  "peerDependencies": {
```
pls add this back in

mojoaxel replied:
@notwaldorf Wow! You actually reviewed this at #jsconfeu ! :tada: Thx. I'll try to have a look tonight.

mojoaxel commented:
@notwaldorf Yo! Check this out! ;-)
(I was a little side-tracked from @fhinkel's talk - hope it works!)

notwaldorf commented:
Awesome! YAY! Thanks!

mojoaxel commented:
:tada: :balloon: !

KernelDeimos commented:
I just cloned the latest changes and received a new error :D (sorry, my computer is probably perfectly screwed up for this, lol)

Any idea what might be causing this?
![screenshot_2017-05-14_14-16-24](https://cloud.githubusercontent.com/assets/7225168/26036683/768c63e0-38b0-11e7-8efb-71e0befc14d8.png)

notwaldorf commented:
@KernelDeimos What's your node version? It seems like it isn't loving the ES6 keywords.

mojoaxel commented:
> Any idea what might be causing this?

Wow! Please update node.js to > [4.8.3](http://node.green/#ES2015-bindings-let-is-block-scoped)

KernelDeimos commented:
My bad - I thought Ubuntu's repos would have it at least reasonably up-to-date :/ boy was I wrong

Garbee commented:
> I haven't found a single reference online that tells me I can just add something to my sources file... what's up with that?

The only stable way to get a newer version without root access is [Node Version Manager](https://github.com/creationix/nvm). It installs node versions into your user folder with your permissions. The os-level update from an external repository will **need** root access since in those cases you are editing your global software sources list and installing the new version.

Generally running trusted scripts as root isn't a problem. You can always download the script and look it over before running it to see what it is doing. If you don't have root access (which is fairly uncommon, especially on workstation machines) then you'll need to go the NVM route.
